app/models/runs_user_reducers.rb

```diff
@@ -0,0 +1,82 @@
+class RunsUserReducers
+  class ReductionConflict < StandardError; end
+
+  attr_reader :reducible, :reducers
+
+  def initialize(reducible, reducers)
+    @reducible = reducible
+    @reducers = reducers
+  end
+
+  def reduce(user_id, extract_ids=[], and_check_rules: false)
+    return [] unless reducers&.present?
+    return [] unless user_id
+    retries ||= 2
+
+    extract_query = prepare_extract_query(user_id)
+    extracts = FetchExtractsByUser.for(reducers).extracts(extract_query, extract_ids)
+
+    reduction_filter = { reducible_id: reducible.id, reducible_type: reducible.class.to_s, user_id: user_id }
+    reduction_fetcher = ReductionFetcher.new(reduction_filter)
+
+    # prefetch all reductions to avoid race conditions with optimistic locking
+    if reducers.any?{ |reducer| reducer.running_reduction? }
+      reduction_fetcher.load!
+    end
+
+    new_reductions = reducers.map do |reducer|
+      reducer.process(
+        extracts, 
+        reduction_fetcher.for!(reducer.topic), 
+        relevant_reductions(extracts, reducer)
+      )
+    end.flatten
+
+    persist_reductions(new_reductions)
+
+    if reducible.is_a?(Workflow) && and_check_rules && new_reductions.present?
+      check_rules(CheckUserRulesWorker, user_id)
+    end
+
+    new_reductions
+  rescue ActiveRecord::StaleObjectError
+    retry unless (retries-=1).zero?
+    raise ReductionConflict, "Object version mismatch"
+  rescue ActiveRecord::RecordNotUnique, PG::UniqueViolation
+    retry unless (retries-=1).zero?
+    raise ReductionConflict, "Transient uniqueness violation"
+  end
+
+  def persist_reductions(reductions)
+    ActiveRecord::Base.transaction do
+      reductions.each do |reduction|
+        reduction.save! unless reduction.user_id.nil?
+      end
+    end
+  end
+  
+  def relevant_reductions(extracts, reducer)
```

Comment:
I'm a little confused by this method name / intention. Specifically, how in the `RunsUserReducers` class, i'm looking up SubjectReductions for this reducible which seems counter intuitive (same for `RunsSubjectReducers`). 

I'm not sure i'm following how the results of this lookup are then used in the `reducer.process` to add the relevant reductions. Would you mind clarifying how this works for me?

Reply:
"Relevant reductions" couple User and Subject reductions for aggregation purposes. Used in TESS right now to send subject difficulties to a user reducer to calculate user skill, and to send user skills to subject reducer to calculate subject probabilities. This will be necessary for any future "user skill" or leveling-style setups.

This should absolutely be documented in the main Caesar docs, that's my bad. I'll open an issue right now.

Reply:
Can we please update the code to better name why / what this is doing? @zwolf maybe add some suggested changes for naming and comments here.

Reply:
Also - it looks like every reducer will load the relevant reductions before passing them to `.process`, correct? 

It sounds like they are only used in TESS which seems wasteful. Instead could we reflect on a reducer setting to only pay this cost when needed and avoid it for every other reducer. Seems pointless to load data we don't use / through away

Reply:
Relevant reductions are determined by a `[subject/user]_reducer_keys` config key on the reducer. So unless it was explicitly set up with that key (via console), it's only wasteful in the sense that its getting an empty relation back from AR. It's not loading any reductions where that config isn't set. 

Could certainly check for that key and only ask for the relevant reductions if it's set, saving a query. Not sure if that's in scope for this PR, though.

Reply:
Ah - thanks for clarifying, that query will still hit the db to filter out the known Reductions that match the filter predicates, however it looks like there is a covering index on those queries so it's probably a cheap(ish) query, not as cheap as no query though ;)

So, to optimize we could avoid this query by reflecting on the `reducer.subject_reducer_keys` attribute before mapping the subject_ids. 
E.g 
```
if reducer.subject_reducer_keys
  subject_ids = extracts.map(&:subject_id)
  SubjectReduction.where(subject_id: subject_ids, reducible: reducible, reducer_key: reducer.subject_reducer_keys)
else
  []
end
```

I'm not convinced we need it though, based on your comments and the covering indexes.